metricflow/test/integration/test_cases/itest_metrics.yaml

```diff
@@ -1043,3 +1043,77 @@ integration_test:
     ) b
     ON a.metric_time__week = b.metric_time__week
       AND a.metric_time__month = b.metric_time__month
+---
+integration_test:
+  name: simple_query_with_date_part
+  description: Test query using date_part
+  model: SIMPLE_MODEL
+  metrics: ["bookings"]
+  group_by_objs: [{"name": "metric_time", "date_part": "year"}]
+  check_query: |
+    SELECT
+      SUM(1) AS bookings
+      , {{ render_extract("ds", DatePart.YEAR) }} AS metric_time__extract_year
+    FROM {{ source_schema }}.fct_bookings
+    GROUP BY {{ render_extract("ds", DatePart.YEAR) }};
+---
+integration_test:
+  name: simple_query_with_multiple_date_parts
+  description: Test query using multiple date_parts
+  model: SIMPLE_MODEL
+  metrics: ["bookings"]
+  group_by_objs: [
+    {"name": "metric_time", "date_part": "quarter"},
+    {"name": "metric_time", "date_part": "dow"},
+    {"name": "metric_time", "date_part": "doy"},
+    {"name": "metric_time", "date_part": "day"},
+    {"name": "metric_time", "date_part": "week"},
+  ]
+  check_query: |
+    SELECT
+      SUM(1) AS bookings
+      , {{ render_extract("ds", DatePart.QUARTER) }} AS metric_time__extract_quarter
+      , {{ render_extract("ds", DatePart.DOW) }} AS metric_time__extract_dow
+      , {{ render_extract("ds", DatePart.DOY) }} AS metric_time__extract_doy
+      , {{ render_extract("ds", DatePart.DAY) }} AS metric_time__extract_day
+      , {{ render_extract("ds", DatePart.WEEK) }} AS metric_time__extract_week
+    FROM {{ source_schema }}.fct_bookings
+    GROUP BY
+      {{ render_extract("ds", DatePart.QUARTER) }}
+      , {{ render_extract("ds", DatePart.DOW) }}
+      , {{ render_extract("ds", DatePart.DOY) }}
+      , {{ render_extract("ds", DatePart.DAY) }}
+      , {{ render_extract("ds", DatePart.WEEK) }};
+---
+integration_test:
+  name: derived_metric_offset_window_and_date_part
+  description: Tests a derived metric offset query with window and date_part
+  model: SIMPLE_MODEL
+  metrics: ["bookings_5_day_lag"]
+  group_by_objs: [{"name": "metric_time", "date_part": "month"}]
+  check_query: |
+    SELECT
+      {{ render_extract("a.ds", DatePart.MONTH) }} AS metric_time__extract_month
+      , SUM(b.bookings_5_day_lag) AS bookings_5_day_lag
+    FROM {{ mf_time_spine_source }} a
+    INNER JOIN (
+      SELECT
+        ds AS metric_time__day
+        , 1 AS bookings_5_day_lag
+      FROM {{ source_schema }}.fct_bookings
+    ) b
+    ON {{ render_date_sub("a", "ds", 5, TimeGranularity.DAY) }} = b.metric_time__day
+    GROUP BY metric_time__extract_month
+---
+integration_test:
+  name: date_part_overrides_granularity
+  description: Test query using date_part with incompatible granularity; should override granularity
```

Comment:
Wait, really? Shouldn't this just be an error? I'd rather throw an exception here, it's hard to know what output the user is expecting.

Like the only time I can think of where I would use this is I want to do some comparison for monthly metrics based on which day of the week the month starts on, in which case making these daily is incorrect. Or I might just be doing something silly, in which case an error is appropriate.

Reply:
Oooh interesting. I was not imagining we would want to support querying a date part smaller than the granularity requested. In that case I figured we would just error (which is what I initially implemented). But then I ran into some issues with setting a default granularity based on what was defined in the metric, and ended up deciding to just override what was requested since, if the granularity is >= the date part, changing the granularity should not impact the date part.

Reply:
IMO, we should not support changing the granularity and then requesting a date part on that new granularity. The integration we are building for currently only allows one or the other. If we were to support both in combination, it would greatly complicate the code & SQL - we would need a new column for each _combination_ of granularity & date part in the source dataset. It seems like a very unlikely edge case, but one that we could later decide to support if customers ask for it.
I'm now leaning toward this option: disallow querying `granularity` and `date_part` on the same time dimension. Instead, if `date_part` is requested, we assume the default granularity from the metric (or time dimension if no metric is requested).
WDYT?

Reply:
All of this would be so much easier if it didn't have to exist inside MetricFlow itself, but I don't think we have that option at all.

I have to be honest, I'm not entirely sure why we need all of the combinations of all of the granularities in the rendered SQL. I mean we clearly don't, because the optimized SQL doesn't include them all, and it's not like the optimizer has more information than the dataflow plan builder. However, fixing *that* bit of silliness is way beyond the scope of this PR.

Since our core entry-point only allows one or the other I think it's fine to have partial support. We do have some interfaces where users must specify granularity (e.g., the `TimeDimension` query parameter object), so I think we can just add validation that the granularity matches the default, and we can infer grain on the `Dimension()` parameter objects.

Just to make sure I get it, we'll then have:

1. Inference for the default granularity if extract is used without granularity specified
2. Validation that granularity matches the default for cases where both are provided

Ideally we'd put a docblock on a validation function explaining why this exists, because I will definitely forget all about this conversation by the time someone gets annoyed by the limitation.

Reply:
Does that mean we should error if a granularity is requested that does not match the default? Even if we _could_ technically get the requested date part with that granularity?

Reply:
I don't know? Let's chat about this tomorrow, I think we're well into the stage where talking will be a lot faster.